src/Codegen/Builders/CompositeBuilder.hack

```diff
@@ -0,0 +1,88 @@
+namespace Slack\GraphQL\Codegen;
+
+use namespace HH\Lib\{Vec, Keyset};
+use type Facebook\HackCodegen\{
+    CodegenClass,
+    CodegenMethod,
+    HackBuilderValues,
+    HackCodegenFactory,
+    CodegenClassConstant,
+};
+
+
+/**
+ * Builder for GraphQL objects and interfaces (has a subclass for each).
+ *
+ * The annotated Hack type should be either a class, interface, or shape.
+ */
+abstract class CompositeBuilder<TField as IFieldBuilder>
+    extends OutputTypeBuilder<\Slack\GraphQL\__Private\CompositeType> {
+
+    public function __construct(
+        \Slack\GraphQL\__Private\CompositeType $type_info,
+        string $hack_type,
+        protected vec<TField> $fields,
+    ) {
+        parent::__construct($type_info, $hack_type);
+    }
+
+    final public static function fromTypeAlias<T>(
```

Comment:
Should this live on `ObjectBuilder`? Or will we support building interfaces from shapes?

Reply:
Good catch, I don't think we can support shape-backed interfaces because we would have no way to map a value to the correct subtype at runtime (the exact thing I'm adding in this PR for classes). I'll move it.